### ISSUE TITLE
Prod IT fails to index lm5 project snapshots

### ISSUE THREAD

**Issue by achave11-ucsc** [DataBiosphere/azul#6081]
https://gitlab.azul.data.humancellatlas.org/ucsc/azul/-/jobs/55501

source.spec `tdr:datarepo-ac5f1124:snapshot/lungmap_prod_1977dc4784144263a8706b0f207d8ab3__20240206_20240222_lm5` fails with:
```
Traceback (most recent call last):
  File "/var/task/azul/indexer/index_controller.py", line 180, in contribute
    contributions, replicas = self.transform(catalog, notification, delete)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/index_controller.py", line 232, in transform
    bundle = service.fetch_bundle(catalog, bundle_fqid)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/index_service.py", line 194, in fetch_bundle
    return plugin.fetch_bundle(bundle_fqid)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr.py", line 212, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 362, in _emulate_bundle
    entities, root_entities, links_jsons = self._stitch_bundles(bundle)
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 438, in _stitch_bundles
    upstream = self._find_upstream_bundles(source, all_dangling_inputs)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 564, in _find_upstream_bundles
    require(not missing,
  File "/var/task/azul/__init__.py", line 1525, in require
    reject(not condition, *args, exception=exception)
  File "/var/task/azul/__init__.py", line 1544, in reject
    raise exception(*args)
azul.RequirementError: Dangling inputs not found in any bundle: {'bc8ad5d2-c9c7-432f-a346-59213f667bea'}
```

source.spec `tdr:datarepo-aa408857:snapshot/lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240222_lm5` fails with:
```
[ERROR] RequirementError: Dangling inputs not found in any bundle: {'bc8ad5d2-c9c7-432f-a346-59213f667bea'}
Traceback (most recent call last):
  File "/var/task/azul/chalice.py", line 166, in patched_event_source_handler
    return old_handler(self_, event, context)
  File "/var/task/chalice/app.py", line 1756, in __call__
    return self.handler(event_obj)
  File "/var/task/app.py", line 337, in contribute_retry
    app.index_controller.contribute(event, retry=True)
  File "/var/task/azul/indexer/index_controller.py", line 180, in contribute
    contributions, replicas = self.transform(catalog, notification, delete)
  File "/var/task/azul/indexer/index_controller.py", line 232, in transform
    bundle = service.fetch_bundle(catalog, bundle_fqid)
  File "/var/task/azul/indexer/index_service.py", line 194, in fetch_bundle
    return plugin.fetch_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr.py", line 212, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 362, in _emulate_bundle
    entities, root_entities, links_jsons = self._stitch_bundles(bundle)
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 438, in _stitch_bundles
    upstream = self._find_upstream_bundles(source, all_dangling_inputs)
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 564, in _find_upstream_bundles
    require(not missing,
  File "/var/task/azul/__init__.py", line 1525, in require
    reject(not condition, *args, exception=exception)
  File "/var/task/azul/__init__.py", line 1544, in reject
    raise exception(*args)
```

**Comment by hannes-ucsc** [comment]
At least one of the new snapshots is buggy. It has dangling inputs. A dangling input is basically an edge in a subgraph (links.json) that points to a non-existing vertex. To address this we had to revert the changes for #6068.

**Comment by dsotirho-ucsc** [comment]
@hannes-ucsc: "A dangling input is a process input that is declared in `links.json` but doesn't actually exist."

**Comment by dsotirho-ucsc** [comment]
The Staging Import Validator (https://github.com/DataBiosphere/hca-import-validation) could be useful here.

**Comment by bvizzier-ucsc** [comment]
It appears that the upload of one of the lm5 datasets was incomplete.  The dataset has been uploaded again and a new snapshot has been requested.   Once that is done, we will schedule a new index run.

**Comment by ebardes** [comment]
I'm trying to run the hca-import-validator.  It's looking for a file called staging_area.json.  Any idea what should be in there?

**Comment by hannes-ucsc** [comment]
> I'm trying to run the hca-import-validator. It's looking for a file called staging_area.json. Any idea what should be in there?

https://github.com/HumanCellAtlas/dcp2/blob/main/docs/dcp2_system_design.rst#staging-area-properties

**Comment by achave11-ucsc** [comment]
The snapshot, `tdr:datarepo-aa408857:snapshot/lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240222_lm5` is still malformed but due to different reasons: 

```
Traceback (most recent call last):
  File "/var/task/azul/indexer/index_controller.py", line 180, in contribute
    contributions, replicas = self.transform(catalog, notification, delete)
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/index_controller.py", line 232, in transform
    bundle = service.fetch_bundle(catalog, bundle_fqid)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/index_service.py", line 194, in fetch_bundle
    return plugin.fetch_bundle(bundle_fqid)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr.py", line 212, in fetch_bundle
    bundle = self._emulate_bundle(bundle_fqid)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 392, in _emulate_bundle
    raise e
  File "/var/lang/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 517, in _retrieve_entities
    rows = self._query_latest_version(source, query, group_by=pk_column)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 347, in _query_latest_version
    return [self._choose_one_version(source, group) for _, group in groups]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 347, in <listcomp>
    return [self._choose_one_version(source, group) for _, group in groups]
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/plugins/repository/tdr_hca/__init__.py", line 354, in _choose_one_version
    return one(versioned_items)
           ^^^^^^^^^^^^^^^^^^^^
  File "/opt/python/more_itertools/more.py", line 554, in one
    raise too_long or ValueError(msg)
ValueError: Expected exactly one item in iterable, but got Row(('0c422f39-b2b1-4cf9-9096-b48517c72191',
 datetime.datetime(2022, 12, 15, 12, 39, 12, 341323, tzinfo=datetime.timezone.utc), 862, '{"biomaterial_core":
{"biomaterial_id":"0c422f39-b2b1-4cf9-9096-b48517c72191","biomaterial_name":"ACD5 Cells","describedBy":
"https://schema.humancellatlas.org/core/biomaterial/8.2.0/biomaterial_core","ncbi_taxon_id":
[9606]},"describedBy":"https://schema.humancellatlas.org/type/biomaterial/13.3.0/cell_suspension","genus_species":
[{"describedBy":"https://schema.humancellatlas.org/module/ontology/5.3.5/species_ontology","ontology":"NCBITaxon:
9606","ontology_label":"Homo sapiens","text":"Homo sapiens"}],"provenance":{"describedBy":
"https://schema.humancellatlas.org/system/1.1.0/provenance","document_id":
"0c422f39-b2b1-4cf9-9096-b48517c72191","submission_date":"2022-11-15T21:33:40.760875Z","submitter_id":"Shuyang.Zhao.cchmc@lungmap.net",
"update_date":"2022-12-15T12:39:12.341323Z","updater_id":"josh.fortriede.cchmc@lungmap.net"},"schema_type":
"biomaterial"}', 'biomaterial'), {'cell_suspension_id': 0, 'version': 1, 'content_size': 2, 'content': 3, 'schema_type': 4}),
 Row(('0c422f39-b2b1-4cf9-9096-b48517c72191', datetime.datetime(2022, 12, 15, 12, 39, 12, 341323, tzinfo=datetime.timezone.utc), 
862, '{"biomaterial_core":{"biomaterial_id":"0c422f39-b2b1-4cf9-9096-b48517c72191","biomaterial_name":
"ACD5 Cells","describedBy":"https://schema.humancellatlas.org/core/biomaterial/8.2.0/biomaterial_core",
"ncbi_taxon_id":[9606]},"describedBy":"https://schema.humancellatlas.org/type/biomaterial/13.3.0/
cell_suspension","genus_species":[{"describedBy":"https://schema.humancellatlas.org/module/ontology/5.3.5/
species_ontology","ontology":"NCBITaxon:9606","ontology_label":"Homo sapiens","text":"Homo sapiens"}],
"provenance":{"describedBy":"https://schema.humancellatlas.org/system/1.1.0/provenance","document_id":
"0c422f39-b2b1-4cf9-9096-b48517c72191","submission_date":"2022-11-15T21:33:40.760875Z","submitter_id":"Shuyang.Zhao.cchmc@lungmap.net",
"update_date":"2022-12-15T12:39:12.341323Z","updater_id":"josh.fortriede.cchmc@lungmap.net"},"schema_type":
"biomaterial"}', 'biomaterial'), {'cell_suspension_id': 0, 'version': 1, 'content_size': 2, 'content': 3, 'schema_type': 4}),
 and perhaps more.
```

**Comment by JoshuaFortriede** [comment]
Parsing the error message, I see two "Row" sections. These look to be tuples with the following elements: UUID, Timestamp, length of content, content (JSON), type.  Both of these Row sections are exactly the same. 

I do not know what this means. Any ideas what is the actual underlying issue? Is it a problem with the JSON or something else.

**Comment by hannes-ucsc** [comment]
1) The original description reports two snapshots as being defective. We have only received an update for one of them.

2) The one we haven't received an update for, is still presenting defective, but with a different symptom. I need to talk to my team in order to determine whether the list of symptoms given originally and yesterday by us was exhaustive. For now please assume that all symptoms apply and that there may be more defects.

3) @achave11-ucsc is not presenting any evidence for the updated snapshot. I take that to mean that the updated snapshot is not defective.

4) Regarding the most recent error

> Parsing the error message, I see two "Row" sections. These look to be tuples with the following elements: UUID, Timestamp, length of content, content (JSON), type. Both of these Row sections are exactly the same.

That observation is correct. Indeed, there are duplicate rows in the snapshot. As far as I can tell, every cell suspension version is listed four times (https://data.terra.bio/snapshots/cf79b25c-3034-4cd0-980f-418d312d4f29/data):

<img width="1118" alt="image" src="https://github.com/DataBiosphere/azul/assets/5143256/a5cdfccd-745e-47b5-a80e-8b2cab4a0ac5">

<p/>

> I do not know what this means. Any ideas what is the actual underlying issue? Is it a problem with the JSON or something else.

I don't think it is possible for a staging area to contain four identical copies of a versioned entity because the ID and version are part of the object name. My best guess is that this is a bug in the importer or TDR.

**Comment by ebardes** [comment]
![image](https://github.com/DataBiosphere/azul/assets/583328/1c766c64-cc0d-4291-b04a-24047a700299)
This is from our graph display tool. There are four processes from a single suspension.

**Comment by ebardes** [comment]
It appears they used different four sequencing technologies.

**Comment by JoshuaFortriede** [comment]
For this cell suspension, ACD5 Cells, the library prep was for multiome sequencing. The result being 4 sets of sequencing files. Two sets being for snATAC-Seq and two sets being for snRNA-Seq. @ebardes can confirm, but I do not believe we would have exported out the ACD5 cells "node" 4 times into the GCP bucket, so I am unsure why the staging area would have it four times.

**Comment by hannes-ucsc** [comment]
> so I am unsure why the staging area would have it four times.

It can't. I checked the [spec](https://github.com/HumanCellAtlas/dcp2/blob/main/docs/dcp2_system_design.rst#object-naming): the entity ID and version are the only components of the object name, and there can only be one object for any given name.

The two columns for entity ID (`cell_suspension_id` in this case) and `version` are the sole primary key columns, so the fact that more than one row have the same value for the primary key is in violation of the [spec](https://github.com/HumanCellAtlas/dcp2/blob/main/docs/dcp2_system_design.rst#tdr-datasets).

**Comment by hannes-ucsc** [comment]
But again, it looks like lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240222_lm5 has two defects, the duplicate cell suspension rows and dangling links.

**Comment by hannes-ucsc** [comment]
I wrote earlier:

> The one we haven't received an update for, is still presenting defective, but with a different symptom. I need to talk to my team in order to determine whether the list of symptoms given originally and yesterday by us was exhaustive. For now please assume that all symptoms apply and that there may be more defects.

I now know why the same snapshot represented with different symptoms last week and yesterday: It is because the symptoms were detected during an integration test that uses a random subset of bundles from each snapshot. Different bundles can have different defects and the IT may not catch all of them. A failing IT usually indicates that there is no point in performing a full reindex because we already know that it would fail. In this case though, I now asked my team to perform a full reindex anyways, so that we get an exhaustive list of defects. I will report back in a couple of hours.

For now, we already know that lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240222_lm5 is afflicted with two issues, duplicate cell suspension rows and dangling links. As far as I can tell, the former is a bug in TDR. If that is true, the Broad Institute needs to be involved in the fix. The latter is a bug in the staging area so the fix would have to come from the LungMAP team and then Broad would have to cut an updated snapshot.

**Comment by ebardes** [comment]
Do you have a UUID from one of the dangling links that might aid my investigation? I'm asking because I am struggling to find a linked input, output, process or protocol that doesn't have a matching object anywhere in our database for any submission past, present or future.

**Comment by nsalomonis** [comment]
Can we exclude the dataset(s) with issues and proceed with the others?

**Comment by hannes-ucsc** [comment]
> Can we exclude the dataset(s) with issues and proceed with the others?

Yes. We'll try to get that done by tomorrow morning. Compared to lm4, it'll just be that one lm6 snapshot.

I also have updated information on the defective lm5 snapshots and an answer to @ebardes question which I will post shortly.

**Comment by hannes-ucsc** [comment]
Despite having been updated, `lungmap_prod_1977dc4784144263a8706b0f207d8ab3__20240206_20240416_lm5` still has the dangling input. The UUID of the dangling input is `bc8ad5d2-c9c7-432f-a346-59213f667bea`. A dangling input is an entity referenced by a `links.json` that is not accounted for. In this case, there is only one `links.json` in the snapshot. Looking at that `links.json`, it becomes evident that the dangling input is a sequence file. There is a corresponding row in the `sequence_file` table but the sequence file also needs to be described as an output of another process in one of the `links.json` files from the same snapshot, either the same `links.json` file that declares the input or another one. As I wrote above, there is only one `links.json` file and the UUID `bc8ad5d2-c9c7-432f-a346-59213f667bea` only occurs once, as an input, but not as an output. The other sequence files do appear to be properly described as an output of another process that derives them from, say, a cell suspension. It is impossible for a sequence file to be created out of thin air so I think Azul's assertion, the failing assertion in this case, is justified. Without it, the sequence file's provenance would be missing.

My team made a mistake in attributing a dangling input to snapshot `lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240222_lm5`. There is no dangling input in that snapshot. In fact, the defect we found yesterday, the duplicate cell suspension rows, is the only type of defect in that snapshot, and its been like that from the beginning when we first created the issue, we just failed to correctly account for it in the description of this issue.

**Comment by nsalomonis** [comment]
Thank you for clarifying. Let me know when the update tomorrow morning is pushed. I believe LM6 is one of our low priority datasets with the LM5 as two high priority with the larger dataset as the main priority. Ideally we would have the LM5 major dataset included (with dozens of input files and a much larger graph).

**Comment by hannes-ucsc** [comment]
Both `lm5` snapshots are defective so they won't be included in the `lm6` release, unfortunately.

**Comment by hannes-ucsc** [comment]
https://github.com/DataBiosphere/azul/issues/6186#issuecomment-2074030009

> We were only able to add `lungmap_prod_6511b041b11e4ccf85932b40148c437e__20240326_20240326_lm6` because the other snapshots slated for inclusion in `lm6` (the ones originally slated for `lm5`) have [defects](https://github.com/DataBiosphere/azul/issues/6081#issuecomment-2073690809). The [spreadsheet](https://docs.google.com/spreadsheets/d/1yYiB3TrX0wvLCHvtWY1GTZZcAf5MmMREgn52OKCHsqs/edit#gid=1910294292) is up-to-date. The `lm6` catalog can be previewed here: https://data-browser.lungmap.net/projects?catalog=lm6

**Comment by ncalvanese1** [comment]
@hannes-ucsc -- Thank you for the detail here. To summarize it sounds like:
- **lungmap_prod_1977dc4784144263a8706b0f207d8ab3__20240206_20240416_lm5** has a dangling input that requires the links.json file to be updated to ensure the sequence file identified by bc8ad5d2-c9c7-432f-a346-59213f667bea exists as an output of a process in addition to an input. @ebardes -- I believe this one will require a resubmission from you all for me to import the corrected data into TDR. 
- **lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240222_lm5** has true duplicates present in the cell_suspension table in the TDR snapshot. I've confirmed this is **NOT** the case in the underlying TDR dataset. So there appears to be something in the TDR snapshotting process that is causing this. I am investigating that point further.

**Comment by JoshuaFortriede** [comment]
@hannes-ucsc, I believe we have fixed the dangling input, and @ebardes will be re-uploading.

**Comment by ncalvanese1** [comment]
The duplication issue with project fdadee7e-2097-45d5-bf81-cc280bd8348e has been resolved and a new snapshot has been issued and added to our tracking sheet (lungmap_prod_fdadee7e209745d5bf81cc280bd8348e__20240206_20240510_lm5)

**Comment by ncalvanese1** [comment]
A replacement snapshot for 1977dc47-8414-4263-a870-6b0f207d8ab3 was issued and added to our tracking sheet as well (lungmap_prod_1977dc4784144263a8706b0f207d8ab3__20240206_20240510_lm5).

**Comment by bvizzier-ucsc** [comment]
lm7 should have addressed this issue.

**Comment by nsalomonis** [comment]
We would like this online to report before June 11th, for LungMAP steering committee update.Best,NathanOn May 15, 2024, at 6:35 PM, Ben Vizzier ***@***.***> wrote:﻿
lm7 should have addressed this issue.

**Comment by hannes-ucsc** [comment]
As far as Azul is concerned, the removal of the defective snapshots addressed this issue. Replacements for the removed snapshots have been provided for inclusion in lm7, but the replacements are also defective. See https://github.com/DataBiosphere/azul/issues/6249 for details. The discussion of the remaining defects should continue there.

> We would like this online to report before June 11th

In order to meet that deadline we would need to receive functional snapshots by 6/3/2024 so that we can complete #6249.
